RFC-0004-support-distributed-procedure.md

```diff
@@ -0,0 +1,397 @@
+
+Add a new type in `QueryType` called `CALL_DISTRIBUTED_PROCEDURE(10)` for `call distributed procedure` statement, it's different from the query type 'DATA-DEFINITION' originally set for `call procedure` statement.
+
+In the preparer phase, when encountering a call statement, utilize the injected `IProcedureRegistry` to determine whether the called procedure is a `DistributedProcedure`. If so, associate its query type to `CALL_DISTRIBUTED_PROCEDURE`.
+
+In this way, in `LocalDispatchQueryFactory` we can choose `SqlQueryExecutionFactory` to generate the query execution for `call distributed procedure` statement. This allows for the utilization of distributed data scanning, writing, and unified committing mechanism during actual execution.
+
+#### 3. Processing logic during analyzing and logical planning phase
+
+Add a new plan node type `CallDistributedProcedureNode` for `call distributed procedure` statement, which is defined as follows:
+
+```java
+public class CallDistributedProcedureNode
+            extends InternalPlanNode
+{
+    private final PlanNode source;
+    private final Optional<CallDistributedProcedureTarget> target;
+    private final VariableReferenceExpression rowCountVariable;
+    private final VariableReferenceExpression fragmentVariable;
+    private final VariableReferenceExpression tableCommitContextVariable;
+    private final List<VariableReferenceExpression> columns;
+    private final List<String> columnNames;
+    private final Set<VariableReferenceExpression> notNullColumnVariables;
+    private final Optional<PartitioningScheme> partitioningScheme;
+    private final List<VariableReferenceExpression> outputs;
+
+    ......
+}
+```
+
+At present, it can be regarded as a simplified version of `TableWriterNode`, and holds `CallDistributedProcedureTarget`, a subclass of `WriterTarget` used to maintain information about the procedure and the call parameters:
+
+```java
+public static class CallDistributedProcedureTarget
+            extends WriterTarget
+{
+    private final QualifiedObjectName procedureName;
+    private final Object[] procedureArguments;
+    private final Optional<TableHandle> sourceHandle;
+    private final SchemaTableName schemaTableName;
+
+    .....
+}
+```
+
+During the analyzing and logical planning phase, the `call distributed procedure` statement will be planned into a logical plan with the following shape:
+
+```text
+TableScanNode -> FilterNode -> CallDistributedProcedureNode -> TableFinishNode -> OutputNode
+```
+
+Among them, `TableScanNode -> FilterNode` defines the data to be processed. It's based on the target table determined by `schema` and `table_name` in the parameters, as well as the possibly existing filter conditions.
+
+The `CallDistributedProcedureNode -> TableFinishNode` structure is similar to the `TableWriterNode -> TableFinishNode` structure, used to perform distributed data manipulation and final unified submission behavior.
```

Comment:
Would be great to give more details on what row/control messages are exchanged between the CallDistricutedProcedureNode and TableFinishNode. e.g. TableWriteNode provides protocol of PartitionUpdates and CommitContext that is shared between the two.

Reply:
Sure, it's a good suggestion. I will supplement this part of the content.